protocol/jsonlines/decoder.go

```diff
@@ -40,17 +41,46 @@ func NewDecoder(r io.Reader) Decoder {
 	return &decoder{r: lr}
 }
 
+// Read and discard everything until the next newline
+func (d *decoder) discardPending() error {
+	for {
+		_, err := d.r.ReadSlice('\n')
+		switch(err) {
+		case io.EOF:
+			return nil
+		case bufio.ErrBufferFull:
+			continue
+		case nil:
+			continue
+		default:
+			return err
+		}
+	}
+}
+
 // Decode decodes the next line in the reader.
 // It does not check JSON for well-formedness before decoding, so in case of
 // error, the structure might be half-filled.
 func (d *decoder) Decode(v interface{}) error {
-	line, isPrefix, err := d.r.ReadLine()
-	if isPrefix {
-		return fmt.Errorf("buffer size exceeded")
-	}
+	var line []byte
+
+	for {
+		chunk, isPrefix, err := d.r.ReadLine()
```

Comment:
Maybe move this code block to a smaller function (e.g. `readLine(r lineReader) error`) to avoid too much nesting and improve readability.